### ISSUE TITLE
`CompositionalMetric` seems not to be logged via `self.log(...)`

### ISSUE THREAD

**Issue by bojobo** [Lightning-AI/torchmetrics#1756]
## 🐛 Bug

When logging a `torchmetrics.CompositionalMetric` via `self.log(...)` the metric should appear in the logs, but in my instance `pytorch_lightning.loggers.WandbLogger` did not report any values.

### To Reproduce

Steps to reproduce the behavior...

<details>
  <summary>Code sample</summary>

This is how I create my loss function (with `psnr_p = 0.5` and `ms_ssim_p = 0.5`):
```py
def create_loss(
        data_scaling,
        l1_p: float = 0.0,
        poisson_p: float = 0.0,
        psnr_p: float = 0.0,
        ssim_p: float = 0.0,
        ms_ssim_p: float = 0.0
) -> Metric:
    total = l1_p + poisson_p + psnr_p + ssim_p + ms_ssim_p
    if not 0.0 < total <= 1.0:
        raise ValueError(f"Sum of l1_p={l1_p}, poisson_p={poisson_p}, psnr_p={psnr_p}, ssim_p={ssim_p}, "
                         f"ms_ssim_p={ms_ssim_p} has to be in ]0.0, 1.0] but is {total}!")

    metrics: Optional[Metric] = None

    if l1_p > 0.0:
        # L1 loss
        scaling, correction = _get_scaling(x1=_zero_epoch[data_scaling]["l1"], x2=_last_epoch[data_scaling]["l1"])
        metrics = MAE(scaling=l1_p * scaling, correction=correction)

    if poisson_p > 0.0:
        # Poisson loss
        scaling, correction = _get_scaling(x1=_zero_epoch[data_scaling]["poisson"],
                                           x2=_last_epoch[data_scaling]["poisson"])
        m = PoissonNLLLoss(scaling=poisson_p * scaling, correction=correction)
        metrics = m if metrics is None else metrics + m

    if psnr_p > 0.0:
        # PSNR loss
        # psnr is a rising metric, therefore inverse the scale
        scaling, correction = _get_scaling(x1=_zero_epoch[data_scaling]["psnr"],
                                           x2=_last_epoch[data_scaling]["psnr"])
        m = PSNR(scaling=psnr_p * scaling, correction=correction)
        metrics = m if metrics is None else metrics + m

    
    # SSIM settings
    if ssim_p > 0.0:
        # ssim is a rising metric, therefore inverse the scale

        # SSIM loss
        scaling, correction = _get_scaling(x1=_zero_epoch[data_scaling]["ssim"],
                                           x2=_last_epoch[data_scaling]["ssim"])
        m = SSIM(scaling=ssim_p * scaling, correction=correction)
        metrics = m if metrics is None else metrics + m

    if ms_ssim_p > 0.0:
        # MS_SSIM loss
        scaling, correction = _get_scaling(x1=_zero_epoch[data_scaling]["ms_ssim"],
                                           x2=_last_epoch[data_scaling]["ms_ssim"])
        m = MultiScaleSSIM(scaling=ms_ssim_p * scaling, correction=correction)
        metrics = m if metrics is None else metrics + m
    return metrics
```
The metrics all inherit from a sub-class of `torchmetrics.Metric` called `_Metric`:
```py
class _Metric(Metric):
    is_differentiable = True
    higher_is_better = True
    full_state_update = False

    def __init__(
            self,
            scaling: float = 1.0,
            correction: float = 0.0,
    ):
        super(_Metric, self).__init__()
        self.add_state("metric", default=torch.tensor(0, dtype=torch.float), dist_reduce_fx="sum")
        self.add_state("total", default=torch.tensor(0, dtype=torch.float), dist_reduce_fx="sum")
        self.scaling = scaling
        self.correction = correction

    def update(
            self,
            preds: torch.Tensor,
            target: torch.Tensor,
            reduction: str = "mean") -> None:
        self.metric += # Here I either call a function from the package piq or from torch.nn.functional in the respective sub-class
        self.total += preds.size()[0]

    def compute(self) -> torch.Tensor:
        scaled = self.scaling * self.metric
        mean_scaled = scaled / self.total
        corrected = mean_scaled + self.correction
        return corrected
```

This happens during `on_{train,  val, test}_step`:
```py
    def _on_step(self, batch, stage) -> Optional[Tensor]:
        lr = batch["lr"]
        preds = self(lr)
        target = batch.get("hr", preds)

        loss = self.loss(preds=preds, target=target)

        if stage == "train":
            self.log(f"{stage}/loss", self.loss, prog_bar=True, batch_size=self.batch_size,
                     on_step=True, on_epoch=False)
            return loss
        else:
            self.log(f"{stage}/loss", self.loss, prog_bar=True, batch_size=self.batch_size,
                     on_step=False, on_epoch=True, sync_dist=True)

            [...]
```

</details>

### Expected behavior

When running `self.log(f"{stage}/loss", self.loss, ...)` the loss should be logged correctly.

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source): 0.11.4 via pip
- Python & PyTorch Version (e.g., 1.0): Python: 3.10.6, PyTorch: 1.13.1 
- Any other relevant information such as OS (e.g., Linux): OS: Ubuntu, PyTorch Lightning version: 1.9.3

### Additional context

Changing the `on_{train, val, test}_step` procedure to the following fixes the bug:
```py
    def _on_step(self, batch, stage) -> Optional[Tensor]:
        lr = batch["lr"]
        preds = self(lr)
        target = batch.get("hr", preds)

        loss = self.loss(preds=preds, target=target)

        if stage == "train":
            self.log(f"{stage}/loss", loss, prog_bar=True, batch_size=self.batch_size,
                     on_step=True, on_epoch=False)
            return loss
        else:
            self.log(f"{stage}/loss", loss, prog_bar=True, batch_size=self.batch_size,
                     on_step=False, on_epoch=True, sync_dist=True)

            [...]
```
